### PR TITLE
ARROW-12102: [C++][Gandiva] Implement new cache for Gandiva focused on a build time policy

### PR DESCRIPTION
**Pull request by jpedroantunes**
Currently gandiva cache is an LRU cache. It would be interesting to change it so that the expressions which take large amount of time to build stay as long as possible in the cache.

So, we need to Implement a new cache for Gandiva focused on this build time policy, and make it configurable so it is possible to define if it will be used the LRU cache or this new one.

### CODE REVIEW

File: cpp/src/gandiva/lower_value_used_cache.h [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+#include <utility>
+
+#include "arrow/util/optional.h"
+#include "gandiva/base_cache.h"
+
+// modified cache to support evict policy of lower value used.
+namespace gandiva {
+// a cache which evicts the lower value used item when it is full
+template <class Key, class Value>
+class LowerValueUsedCache : public BaseCache<Key, Value> {
```
Comment (projjal) [inline cpp/src/gandiva/lower_value_used_cache.h:32]: Just evicting based on the build times seems would be problematic since it might just keep only infrequently used large expressions. You should use a combination of build time and some aging policy.
Also can you change the name. It sounds confusing.

File: cpp/src/gandiva/lower_value_used_cache.h [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+        // check if the value should be inserted on cache, otherwise just return
+        if (value_to_order <= lvu_set_.begin()->first) return;
+
+        // cache is full, evict the least recently used item
+        evict();
+      }
+
+      // insert the new item
+      lvu_set_.insert(std::make_pair(value_to_order, key));
+      map_[key] = std::make_pair(value, lvu_set_.begin());
```
Comment (projjal) [inline cpp/src/gandiva/lower_value_used_cache.h:71]: Why adding lvu_set_.begin()

**Review by projjal — CHANGES_REQUESTED** [review]

File: cpp/src/gandiva/lower_value_used_cache.h (hunk 2 above)
Comment (jpedroantunes) [reply cpp/src/gandiva/lower_value_used_cache.h]: Agree that is not necessary. I will remove it

File: cpp/src/gandiva/base_cache.h [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+#include <list>
+#include <unordered_map>
+#include <utility>
+
+#include "arrow/util/optional.h"
+
+namespace gandiva {
+
+template <class Key, class Value>
+class BaseCache {
```
[the remark this answers is not in the document]
Comment (jpedroantunes) [reply cpp/src/gandiva/base_cache.h:30]: Applied!

File: cpp/src/gandiva/cache.h [hunk 4]
```diff
@@ -31,29 +32,40 @@ int GetCapacity();
...
+GANDIVA_EXPORT
+int GetCacheTypeToUse();
+
 template <class KeyType, typename ValueType>
 class Cache {
  public:
-  explicit Cache(size_t capacity) : cache_(capacity) { LogCacheSize(capacity); }
+  explicit Cache(size_t capacity, int cache_type_to_use) {
+    if (cache_type_to_use == 0) {
+      this->cache_ = new LowerValueUsedCache<KeyType, ValueType>(capacity);
```
[the remark this answers is not in the document]
Comment (jpedroantunes) [reply cpp/src/gandiva/cache.h]: Applied!

File: cpp/src/gandiva/cache.cc [hunk 5]
```diff
@@ -38,6 +42,22 @@ int GetCapacity() {
   return capacity;
 }
 
+int GetCacheTypeToUse() {
+  int cache_type;
+  const char* env_cache_type = std::getenv("GANDIVA_CACHE_TYPE");
```
[the remark this answers is not in the document]
Comment (jpedroantunes) [reply cpp/src/gandiva/cache.cc:47]: I will check the code to better understand what you suggested. If I have any problem, I will reach you

File: cpp/src/gandiva/lower_value_used_cache.h (hunk 1 above)
Comment (jpedroantunes) [reply cpp/src/gandiva/lower_value_used_cache.h:32]: I agree with you!
What I want to discuss here is what would be the best trigger to remove an element according to the build time policy or the recency policy. Do you have any opinion about it?

Comment (projjal) [reply cpp/src/gandiva/lower_value_used_cache.h:32]: Something like https://en.wikipedia.org/wiki/Cache_replacement_policies#LFU_with_dynamic_aging_(LFUDA)

File: cpp/src/gandiva/filter.h [hunk 6]
```diff
@@ -45,6 +45,10 @@ class FilterCacheKey {
 
   bool operator!=(const FilterCacheKey& other) const { return !(*this == other); }
 
+  bool operator<(const FilterCacheKey& other) const {
+    return uniqifier_ < other.uniqifier_;
```
Comment (projjal) [inline cpp/src/gandiva/filter.h:49]: This is not correct I believe. Also why is this needed?

Comment (jpedroantunes) [reply cpp/src/gandiva/filter.h:49]: We needed to implement it because actually, we use this Filter key in a pair structure inside a set on the new Gandiva Cache.
We use this set to keep registers ordered by the elapsed build time.
This is why we need to implement this operator anyway, otherwise we can not use it as I needed in logic.

Comment (projjal) [reply cpp/src/gandiva/filter.h:49]: You should compare on all the properties. uniqifier property is most likely 0 for all expressions

Comment (jpedroantunes) [reply cpp/src/gandiva/filter.h:49]: Not sure if I completely understand your point, but I compared all possible properties for the bool operator<

**Comment by jpedroantunes** [comment]
Will be implemented later
